src/coreclr/System.Private.CoreLib/src/System/Enum.EnumInfo.CoreCLR.cs

```diff
@@ -0,0 +1,89 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Runtime.CompilerServices;
+
+namespace System
+{
+    public abstract partial class Enum
+    {
+        internal sealed unsafe partial class EnumInfo
+        {
+            /// <summary>
+            /// Set by <see cref="EnsureReflectionInfoIsInitialized(RuntimeType, bool)"/>, tracks when
+            /// <see cref="HasFlagsAttribute"/>, <see cref="Values"/> and <see cref="Names"/> have been set.
+            /// </summary>
+            private bool _isReflectionInfoInitialized;
+
+            // Same values as the shared EnumInfo, but nullable due to lazy initialization in CoreCLR
+            public ulong[]? Values;
+            public string[]? Names;
+
+            /// <summary>
+            /// The cached managed function pointer to the right JIT allocation helper.
+            /// Same as in RuntimeType.ActivationCache, just copied here to avoid cache conflicts.
+            /// </summary>
+            private delegate*<void*, object> _pfnAllocator;
+
+            /// <summary>
+            /// Initializes the reflection info (<see cref="HasFlagsAttribute"/>, <see cref="Values"/> and <see cref="Names"/>).
+            /// </summary>
+            /// <param name="enumType">The enum type to use.</param>
+            /// <param name="getNames">Whether or not to also retrieve the enum names.</param>
+            public void EnsureReflectionInfoIsInitialized(RuntimeType enumType, bool getNames)
+            {
+                if (!_isReflectionInfoInitialized || (getNames && Names is null))
+                {
+                    ulong[]? values = null;
+                    string[]? names = null;
+                    RuntimeTypeHandle enumTypeHandle = enumType.TypeHandle;
+
+                    GetEnumValuesAndNames(
+                        new QCallTypeHandle(ref enumTypeHandle),
+                        ObjectHandleOnStack.Create(ref values),
+                        ObjectHandleOnStack.Create(ref names),
+                        getNames ? Interop.BOOL.TRUE : Interop.BOOL.FALSE);
+
+                    bool hasFlagsAttribute = enumType.IsDefined(typeof(FlagsAttribute), inherit: false);
+
+                    HasFlagsAttribute = hasFlagsAttribute;
+                    Values = values!;
```

Comment:
This has race condition. It can overwrite already-initialized Names back to null.

Reply:
Ah, right, I missed this corner case with a thread not requesting names racing against another one. I've added a compare exchange to fix this and a comment explaining the issue here. Left the previous fields lock-free since repeated writes should not cause issues as far as I can tell. Thanks! 🙂